node/engine/chainservice/eth_chainservice.go

```diff
@@ -271,12 +291,63 @@ out:
 			// Due to https://github.com/ethereum/go-ethereum/issues/23845 we can't rely on a long running subscription.
 			// We unsub here and recreate the subscription in the next iteration of the select.
 			sub.Unsubscribe()
-		case chainEvent := <-logs:
+		case naChainEvent := <-logs:
+			// Make sure we care about this event. If we do, process asynchronously
+			for _, topic := range topicsToWatch {
+				if naChainEvent.Topics[0] == topic {
+					go ecs.waitToProcessEvent(naChainEvent, errorChan)
+					break
+				}
+			}
+		}
+	}
+}
+
+func (ecs *EthChainService) waitToProcessEvent(chainEvent ethTypes.Log, errorChan chan<- error) {
+	ecs.logger.Debug().Msgf("event emitted in tx %s (topic: %s)", chainEvent.TxHash.Hex(), chainEvent.Topics[0])
+	initialBlock := big.NewInt(int64(chainEvent.BlockNumber))
+
+	ticker := time.NewTicker(BLOCK_QUERY_DELAY_SECONDS * time.Second)
```

Comment:
In an earlier implementation I created a `ecs.chain.NewHead` subscription that would trigger each time a new block was mined on the network. However, that implementation seemed to be more heavyweight because a single transaction could cause us to generate many equivalent subscriptions. This ticker implementation seems like it will be less of a strain on the node when we have many concurrent events because we just poll once every `BLOCK_QUERY_DELAY_SECONDS`

Reply:
A couple thoughts:
1. Chain events and transactions will be relatively rare events in a nitro network (compared to state updates).
2. Is there a way to have a single `NewHead` subscription that is shared? It would be great to explore this approach before resorting to polling.

Reply:
One approach that would only use a single `NewHead` subscription and eliminate polling:

1. Create a single `newBlocks` channel that is always written to whenever a new block is mined
```
newBlocks := make(chan ethTypes.Header)
sub, err := ecs.chain.SubscribeNewHead(ecs.ctx, newBlocks)
```
2. Create a priority queue (aka heap) to keep track of emitted logs that are waiting to be processed. Can use the golang "container/heap" package for this.
3. Modify our `listenForLogs` loop to check for newBlocks. When a new block is received, we can check the queue and asynchronously trigger the event processor for all events that have enough block confirmations.
```
case naChainEvent := <-logs:
	// When we get a log, add it to the heap
	heap.Push(logsHeap, LogEvent{
		Log:   naChainEvent,
		Block: big.NewInt(int64(naChainEvent.BlockNumber)),
	})

case newBlock := <-newBlocks:
	// When we get a new block, check the logs at the top of the queue.
	// If they're ready to be processed, process them and remove them from the queue.
	for logsHeap.Len() > 0 && newBlock.Number().Cmp(new(big.Int).Add(logsHeap[0].Block, big.NewInt(REQUIRED_CONFIRMATION_BLOCKS))) >= 0 {
		chainEvent := heap.Pop(logsHeap).(LogEvent)
		go ecs.dispatchChainEvents(chainEvent)
	}
}
```

Reply:
Logs are available for the latest run of a PR branch. The logs are under Summary -> Artifacts. For example:
<img width="1251" alt="Screenshot 2023-08-14 at 8 28 27 AM" src="https://github.com/statechannels/go-nitro/assets/1062379/f489c779-d4fe-46f6-9d36-43d9ed4870c9">